include/fiction/algorithms/simulation/sidb/displacement_robustness_domain.hpp

```diff
@@ -0,0 +1,670 @@
+     */
+    std::size_t num_operational_sidb_displacements{0};
+    /**
+     * The number of non-operational SiDB layouts resulting from the given layout by displacements.
+     */
+    std::size_t num_non_operational_sidb_displacements{0};
+};
+
+namespace detail
+{
+
+template <typename Lyt, typename TT>
+class displacement_robustness_domain_impl
+{
+  public:
+    /**
+     * Standard constructor. Initializes the layout, the truth table, the parameters, and the statistics.
+     *
+     * @param lyt SiDB cell-level layout to be evaluated regarding displacement robustness.
+     * @param spec Expected Boolean function of the layout given as a multi-output truth table.
+     * @param ps Parameters for the displacement robustness computation.
+     * @param st Statistics related to the displacement robustness computation.
+     */
+    displacement_robustness_domain_impl(const Lyt& lyt, const std::vector<TT>& spec,
+                                        const displacement_robustness_domain_params<cell<Lyt>>& ps,
+                                        displacement_robustness_domain_stats&                   st) noexcept :
+            layout{lyt},
+            params{ps},
+            stats{st},
+            truth_table{spec},
+            generator(rd())
+    {
+        assert(
+            (is_operational(layout, truth_table, params.operational_params).first == operational_status::OPERATIONAL) &&
+            "The given layout is not a valid SiDB layout for the given Boolean function");
+
+        if (params.analysis_mode ==
+            displacement_robustness_domain_params<cell<Lyt>>::displacement_analysis_mode::RANDOM)
+        {
+            assert(params.percentage_of_analyzed_displaced_layouts >= 0.0 &&
+                   params.percentage_of_analyzed_displaced_layouts <= 1.0 &&
+                   "percentage_of_analyzed_displaced_layouts must be between 0.0 and 1.0");
+        }
+
+        sidbs_of_the_original_layout.reserve(layout.num_cells());
+        layout.foreach_cell([&](const auto& c) { sidbs_of_the_original_layout.push_back(c); });
+    };
+    /**
+     * This function calculates the robustness domain of the SiDB layout based on the provided truth table specification
+     * and displacement robustness computation parameters.
+     */
+    displacement_robustness_domain<Lyt> determine_robustness_domain() noexcept
+    {
+        mockturtle::stopwatch stop{stats.time_total};
+        calculate_all_possible_displacements_for_each_sidb();
+
+        auto layouts = generate_valid_displaced_sidb_layouts();
+
+        if (layouts.empty())
+        {
+            return displacement_robustness_domain<Lyt>{};
+        }
+
+        // Shuffle the layouts vector to have random displaced layouts
+        std::shuffle(layouts.begin(), layouts.end(), generator);
+
+        displacement_robustness_domain<Lyt> domain{};
+
+        std::mutex mutex_to_protect_shared_resources{};
```

Comment:
The name is funny because that's what all mutexes are for :joy: can we make it more descriptive?